### PR TITLE
Trie: Add support for values and prefix nodes

### PR DESCRIPTION
**Pull request by sirspinach**
A prefix key "asdf" can be inserted into the Trie to match all keys that begin with "asdf". `Lookup()` returns the value associated with the most specific match, prioritizing an exact match over the longest prefix.

### CODE REVIEW

File: core/modules/url_filter.cc [hunk 1]
```diff
@@ -141,8 +142,8 @@ CommandResponse UrlFilter::Init(const bess::pb::UrlFilterArg &arg) {
                          std::forward_as_tuple(url.host()),
                          std::forward_as_tuple());
     }
-    Trie &trie = blacklist_.at(url.host());
-    trie.Insert(url.path());
+    Trie<std::tuple<>> &trie = blacklist_.at(url.host());
+    trie.Insert(url.path(), {});
```
Comment (barath) [inline core/modules/url_filter.cc]: Perhaps rewrite these two lines as:
`blacklist_[url.host()].Insert(url.path(), {});`

File: core/utils/trie.h [hunk 2]
```diff
@@ -34,26 +40,47 @@ class Trie {
...
 
   Trie() : root_() {}
-  Trie(const Trie& t) : root_(t.root_) { }
+  Trie(const Trie& t) : root_(t.root_) {}
 
-  // Inserts a string into the trie
-  void Insert(const std::string& key);
+  // Inserts a string into the trie, associating the key
+  // with the value.
+  void Insert(const std::string& key, const T val);
```
Comment (barath) [inline core/utils/trie.h]: You should probably have `val` be passed as a const reference since T is likely not a reference type.  (Here and for other methods.)

File: core/utils/trie_test.cc [hunk 3]
```diff
@@ -1,91 +1,190 @@
...
+  EXPECT_TRUE(trie2.MatchPrefix(""));
+  EXPECT_TRUE(trie2.MatchPrefix("1"));
+
+  EXPECT_TRUE(trie2.Lookup("Hello").first);
+  EXPECT_TRUE(trie2.Lookup("Hello y'all").first);
+  EXPECT_TRUE(trie2.Lookup("Hello World").first);
+
+  EXPECT_EQ(trie2.Lookup("Hello").second, 2);
+  EXPECT_EQ(trie2.Lookup("Hello y'all").second, 2);
+  EXPECT_EQ(trie2.Lookup("Hello World").second, 2);
```
Comment (nrdmao) [inline core/utils/trie_test.cc]: The interesting corner cases that arise with respect to the management of the Trie are inserts and deletes which force the creation and destruction of Nodes and the associated processing of the children array. A test that randomly creates thousands of keys of say between 1 and 8 characters and then randomly removes and inserts keys while ensuring that lookups work as expected might more thoroughly prove out these algorithms.

Comment (sirspinach) [inline core/utils/trie_test.cc]: I haven't added deletion to the Trie yet, but I'll remember to write a test like that when I do.

**Comment by sangjinhan** [comment]
62269c9bb6312f35247579865d895a33
